### PR TITLE
[Datagrid] fix for infinite rerender when `flex` container used as parent.

### LINKED ISSUE #8574: [DataGrid] Container with 100% width and rows overflowing its height, shows jumpy behavior
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

Link to live example:

https://codesandbox.io/s/mui-x-issue-template-forked-e2m06l?file=/src/App.tsx

### Current behavior 😯

DataGrid jumps horizontally

### Expected behavior 🤔

DataGrid should show an horizontal scroll and don't jump
Exactly the same code worked fine with DataGrid v5

### Context 🔦

The same code worked fine with DataGrid v5

### Your environment 🌎

  Binaries:                                                  
    Node: 18.13.0 
    Yarn: 1.22.19
    npm: 6.14.11
  Browsers:
    Chrome: 112.0.5615.50
    Edge: Spartan (44.19041.1266.0), Chromium (112.0.1722.34)
  npmPackages:
    @emotion/react:  11.10.5 
    @emotion/styled:  11.10.5 
    @mui/base:  5.0.0-alpha.123 
    @mui/core-downloads-tracker:  5.11.15 
    @mui/icons-material: ^5.11.11 => 5.11.11 
    @mui/material: ^5.11.15 => 5.11.15 
    @mui/private-theming:  5.11.13 
    @mui/styled-engine:  5.11.11                             
    @mui/styles: ^5.11.13 => 5.11.13                         
    @mui/system:  5.11.15                                    
    @mui/types:  7.2.3                                       
    @mui/utils:  5.11.13                                     
    @mui/x-data-grid: ^6.1.0 => 6.1.0
    @mui/x-data-grid-premium: ^6.1.0 => 6.1.0
    @mui/x-data-grid-pro:  6.1.0
    @mui/x-date-pickers: ^6.1.0 => 6.1.0
    @mui/x-license-pro:  6.0.4
    @types/react:  18.0.28
    react: ^18.2.0 => 18.2.0
    react-dom: ^18.2.0 => 18.2.0

### Order ID or Support key 💳 (optional)

_No response_

### PR DESCRIPTION
**Pull request by yaredtsy**
fixes #8574

**Problems**

The current implementation causes an issue where the addition of a `scrollbar` creates an offset, leading to a resize of the grid. As a result, `needsHorizontalScrollbar` is computed again, and due to the offset, `containerDimensions.width` increases, causing the scrollbar to be removed. The removal of the scrollbar eliminates the offset, causing another resize of the grid, and leading to an infinite resizing loop.

**Fix**

we need to consider the offset created by scrollbar when computing `needsHorizontalScrollbar`

Before: https://codesandbox.io/s/mui-x-issue-template-forked-e2m06l?file=/src/App.tsx
After:  https://codesandbox.io/s/mui-x-issue-template-forked-xbbwsz?file=/src/index.tsx

### CODE REVIEW

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx [hunk 1]
```diff
@@ -600,8 +600,15 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
     return rows;
   };
 
+  let offset = 0;
+
+  const current = (ref as unknown as React.MutableRefObject<HTMLButtonElement | null>).current;
+  if (current) {
+    offset = current.offsetWidth - current.clientWidth;
+  }
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:608]:
```suggestion
  const scrollBarWidth = (rootRef?.current?.offsetWidth || 0) - (rootRef?.current?.clientWidth || 0);
```

To make it clear that we're calculating scrollbar width here?

**Review by cherniavskii — COMMENTED** [review]
Thanks, @yaredtsy!
Looks like a good fix for now!
Do you think you could cover it with a unit test? There's a simplified reproduction example in https://github.com/mui/mui-x/issues/8574#issuecomment-1505368203 that might be helpful.

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx (hunk 1 above)
Comment (yaredtsy) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:608]: I think it's for the scrollbar arrow. should I name it `scrollBarArrowWidth`

image

Comment (yaredtsy) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:608]: > Do you think you could cover it with a unit test?

sure, but I have some questions about it. can I run the layout test on locally or do I have to push it?

Comment (cherniavskii) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:608]: > I think it's for the scrollbar arrow. should I name it scrollBarArrowWidth

I don't think it's a scrollbar arrow width:

> `offsetWidth`
> offsetWidth is a measurement in pixels of the element's CSS width, **including any borders, padding, and vertical scrollbars** (if rendered)
> https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetWidth

> `clientWidth`
> It includes padding but **excludes borders, margins, and vertical scrollbars** (if present).
> https://developer.mozilla.org/en-US/docs/Web/API/Element/clientWidth

In this case, the element doesn't have borders or padding, so `offsetWidth - clientWidth` gives us the vertical scrollbar width, right?

Comment (cherniavskii) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:608]: > sure, but I have some questions about it. can I run the layout test on locally or do I have to push it?

Yes, you can run unit tests in the browser with `yarn test:karma`. To run only a subset of tests - use `it.only(` or `describe.only(`
Let me know if I missed something

Comment (yaredtsy) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx]: > In this case, the element doesn't have borders or padding, so offsetWidth - clientWidth gives us the vertical scrollbar width, right?

yea, you are right. I thought the horizontal scrollbar was causing the `width` to change. I have not realized the width is also affected by `needsHorizontalScrollbar`.

> Let me know if I missed something

I am having trouble coming up with a test. I could not write a failing test. it's kind of tricky b/c I need to check the grid is resizing continuously.

**Comment by cherniavskii** [comment]
I think we can close it now, since the original issue https://github.com/mui/mui-x/issues/8574 was fixed by https://github.com/mui/mui-x/pull/8091

**Comment by cherniavskii** [comment]
@yaredtsy Thanks for working on it though!

**Pull request closed without merging**